src/expression.js

```diff
@@ -482,30 +512,38 @@ pp.parseObj = function(isPattern, refDestructuringErrors) {
       if (!isPattern)
         isGenerator = this.eat(tt.star)
     }
-    this.parsePropertyName(prop)
-    this.parsePropertyValue(prop, isPattern, isGenerator, startPos, startLoc, refDestructuringErrors)
+    this.parsePropertyName(prop, refDestructuringErrors)
+    if (!isPattern && this.options.ecmaVersion >= 8 && !isGenerator && !prop.computed && prop.key.type === "Identifier" && prop.key.name === "async" && (this.type === tt.name || this.type === tt.bracketL) && !this.canInsertSemicolon()) {
```

Comment:
> `this.type === tt.name || this.type === tt.bracketL`

This doesn't seem quite sufficient; according to the [spec](https://tc39.github.io/ecmascript-asyncawait/#prod-AsyncMethod), `async` methods can have arbitrary `PropertyName`s as their name, so the following should probably be allowed:

```javascript
({
  async "m"() {}
  async 42() {}
})
```